### PR TITLE
Stop using zero-bounds on malloc(0)

### PR DESCRIPTION
**Pull request by brooksdavis**

### CODE REVIEW

File: contrib/jemalloc/src/jemalloc.c [hunk 1]
```diff
@@ -2431,7 +2428,7 @@ je_malloc(size_t size) {
 		LOG("core.malloc.exit", "result: %p", ret);
 
 		/* Fastpath success */
-		return BOUND_PTR(ret, zero_size ? 0 : size);
+		return BOUND_PTR(ret, usize);
```
Comment (jrtc27) [inline contrib/jemalloc/src/jemalloc.c:2431]: You don't mean that. usize is only defined under config_stats || config_prof

Comment (brooksdavis) [reply contrib/jemalloc/src/jemalloc.c:2431]: Indeed. Looks like I should just be calling `sz_index2size(ind)` directly. I suppose we should be storing bounded caps in the cache.

Comment (jrtc27) [reply contrib/jemalloc/src/jemalloc.c]: That won't give precise bounds any more even in cases where it's totally fine?

Comment (jrtc27) [reply contrib/jemalloc/src/jemalloc.c]: Agreed

**Review by bsdjhb — APPROVED** [review]
A few suggestions for the second commit log: s/Previouly/Previously/ and s/were ptr was function./where ptr expanded to a function call./

**Comment by brooksdavis** [comment]
> A few suggestions for the second commit log: s/Previouly/Previously/ and s/were ptr was function./where ptr expanded to a function call./

Fixed up the commit message

File: contrib/jemalloc/include/jemalloc/internal/jemalloc_internal_macros.h [hunk 2]
```diff
@@ -115,10 +115,14 @@ JEMALLOC_DIAGNOSTIC_DISABLE_SPURIOUS
 #define	BOUND_PTR(ptr, size)	(ptr)
 #define	ROUND_SIZE(size)		(size)
 #else
-#define	BOUND_PTR(ptr, size)						\
-    ((ptr == NULL) ? NULL :						\
-	 cheri_andperm(cheri_setboundsexact((ptr), (size)),		\
-	     CHERI_PERMS_USERSPACE_DATA & ~CHERI_PERM_SW_VMEM))
+#define	BOUND_PTR(ptr, size) ({						\
+	typeof(ptr) _ptr = (ptr);					\
```
Comment (bsdjhb) [inline contrib/jemalloc/include/jemalloc/internal/jemalloc_internal_macros.h:119]: Probably you could get by with just `void * _ptr` here if you wanted to?

**Review by bsdjhb — APPROVED** [review]

**Comment by jrtc27** [comment]
Could you please split off the statement expression change? That'll want to be backported, but not the behaviour change for zero-sized allocations.

File: contrib/jemalloc/include/jemalloc/internal/jemalloc_internal_macros.h (hunk 2 above)
Comment (jrtc27) [reply contrib/jemalloc/include/jemalloc/internal/jemalloc_internal_macros.h:119]: Then the macro won't be type-preserving like the intrinsics and the non-CHERI version

**Pull request merged**